### PR TITLE
Retry restarting receiver on PubSub service error

### PR DESCRIPTION
**Pull request by emilymye**
### Required for all PRs:

- [X] Signed [CLA](https://influxdata.com/community/cla/).
- [X] Associated README.md updated.
- [X] Has appropriate unit tests.

### CODE REVIEW

File: plugins/inputs/cloud_pubsub/pubsub.go [hunk 1]
```diff
@@ -78,35 +84,37 @@ func (ps *PubSub) Start(ac telegraf.Accumulator) error {
 		return fmt.Errorf(`"project" is required`)
 	}
 
-	cctx, cancel := context.WithCancel(context.Background())
+	ps.sem = make(semaphore, ps.MaxUndeliveredMessages)
+	ps.acc = ac.WithTracking(ps.MaxUndeliveredMessages)
+
+	// Create top-level context with cancel that will be called on Stop().
+	ctx, cancel := context.WithCancel(context.Background())
+	ps.ctx = ctx
```
Comment (danielnelson) [inline plugins/inputs/cloud_pubsub/pubsub.go]: Don't assign the context to the struct, instead pass it down to the goroutines below and save only the cancel function.

File: plugins/inputs/cloud_pubsub/pubsub.go [hunk 2]
```diff
@@ -119,13 +127,46 @@ func (ps *PubSub) Stop() {
...
+
+		log.Printf("I! Waiting %d seconds before attempting to restart receiver...", delay)
+		time.Sleep(time.Duration(delay) * time.Second)
+
+		err = ps.startReceiver()
+	}
+}
+
+func (ps *PubSub) startReceiver() error {
+	cctx, ccancel := context.WithCancel(ps.ctx)
```
Comment (danielnelson) [inline plugins/inputs/cloud_pubsub/pubsub.go]: Move the creation of the cancel function into receiveOnSub, and just pass the context down.

File: plugins/inputs/cloud_pubsub/pubsub.go [hunk 3]
```diff
@@ -119,13 +127,46 @@ func (ps *PubSub) Stop() {
...
 			ps.acc.AddError(fmt.Errorf("unable to add message from subscription %s: %v", ps.sub.ID(), err))
 		}
 	})
-	ps.acc.AddError(fmt.Errorf("receiver for subscription %s exited: %v", ps.sub.ID(), err))
+	if err != nil {
+		ps.acc.AddError(fmt.Errorf("receiver for subscription %s exited: %v", ps.sub.ID(), err))
+	} else {
+		log.Printf("I! subscription pull ended (no error, most likely stopped)")
+	}
+	ccancel()
```
Comment (danielnelson) [inline plugins/inputs/cloud_pubsub/pubsub.go]: I'm confused here, can you explain what this cancel function is doing?  It seems like the operation is over at this point?

File: plugins/inputs/cloud_pubsub/pubsub.go [hunk 4]
```diff
@@ -119,13 +127,46 @@ func (ps *PubSub) Stop() {
...
 }
 
-func (ps *PubSub) subReceive(cctx context.Context) {
+// startReceiver is called within a goroutine and manages keeping a
+// subscription.Receive() up and running while the plugin has not been stopped.
+func (ps *PubSub) receiveWithRetry() {
+	err := ps.startReceiver()
+
+	for err != nil && ps.ctx.Err() == nil {
+		log.Printf("E! Receiver for subscription %s exited with error: %v", ps.sub.ID(), err)
```
Comment (danielnelson) [inline plugins/inputs/cloud_pubsub/pubsub.go]: On all the logging calls, start them with: `E! [inputs.cloud_pubsub]` (or `I!`, `D!`).  We are looking to simplify this in the future but for now this helps keep straight where the log message is coming from.

File: plugins/inputs/cloud_pubsub/pubsub.go (hunk 3 above)
Comment (emilymye) [reply plugins/inputs/cloud_pubsub/pubsub.go]: I'm just making sure the child context I create per subscription receive is canceled. I believe this needs to be done to guarantee that some message handling is finished at the end of the receive (i.e. if receive fails), and it doesn't hurt even if the Google lib already does it. This was previously being done on the top-level cancel during Stop().

**Pull request merged**